cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,639 @@
+
+
+def _plot_density_matrix(mat: np.ndarray) -> None:
+    a, _ = mat.shape
+    num_qubits = int(np.sqrt(a))
+    state_labels = [[0, 1]] * num_qubits
+    kets = []
+    for label in itertools.product(*state_labels):
+        kets.append('|' + str(list(label))[1:-1] + '>')
+    mat_re = np.real(mat)
+    mat_im = np.imag(mat)
+    _matrix_bar_plot(mat_re, r'Real($\rho$)', kets)
+    _matrix_bar_plot(mat_im, r'Imaginary($\rho$)', kets)
+
+
+def _gate_seq_to_mats(gate_seq: Sequence[ops.Gate]) -> np.ndarray:
+    mat_rep = protocols.unitary(gate_seq[0])
+    for gate in gate_seq[1:]:
+        mat_rep = np.dot(protocols.unitary(gate), mat_rep)
+    return mat_rep
+
+
+def _two_qubit_clifford(q_0: devices.GridQubit, q_1: devices.GridQubit,
+                        idx: int, c1: Sequence[Sequence[ops.Gate]],
+                        s1: Sequence[Sequence[ops.Gate]],
+                        s1_x: Sequence[Sequence[ops.Gate]],
+                        s1_y: Sequence[Sequence[ops.Gate]]
+                        ) -> Iterator[ops.OP_TREE]:
+    """Generates a two-qubit Clifford gate.
+
+    An integer (idx) from 0 to 11519 is used to generate a two-qubit Clifford
+    gate which is constructed with single-qubit X and Y rotations and CZ gates.
+    The decomposition of the Cliffords follow those described in the appendix
+    of Barends et al., Nature 508, 500.
+
+    The integer idx is first decomposed into idx_0 (which ranges from 0 to
+    23), idx_1 (ranging from 0 to 23) and idx_2 (ranging from 0 to 19). idx_0
+    and idx_1 determine the two single-qubit rotations which happen at the
+    beginning of all two-qubit Clifford gates. idx_2 determines the
+    subsequent gates in the following:
+
+    a) If idx_2 = 0, do nothing so the Clifford is just two single-qubit
+    Cliffords (total of 24*24 = 576 possibilities).
+
+    b) If idx_2 = 1, perform a CZ, followed by -Y/2 on q_0 and Y/2 on q_1,
+    followed by another CZ, followed by Y/2 on q_0 and -Y/2 on q_1, followed
+    by one more CZ and finally a Y/2 on q_1. The Clifford is then a member of
+    the SWAP-like class (total of 24*24 = 576 possibilities).
+
+    c) If 2 <= idx_2 <= 10, perform a CZ followed by a member of the S_1
+    group on q_0 and a member of the S_1^(Y/2) group on q_1. The Clifford is
+    a member of the CNOT-like class (a total of 3*3*24*24 = 5184 possibilities).
+
+    d) If 11 <= idx_2 <= 19, perform a CZ, followed by Y/2 on q_0 and -X/2 on
+    q_1, followed by another CZ, and finally a member of the S_1^(Y/2) group on
+    q_0 and a member of the S_1^(X/2) group on q_1. The Clifford is a member
+    of the iSWAP-like class (a total of 3*3*24*24 = 5184 possibilities).
+
+    Through the above process, all 11520 members of the two-qubit Clifford
+    group may be generated.
+
+    Args:
+        q_0: The first qubit under test.
+        q_1: The second qubit under test.
+        idx: An integer from 0 to 11519.
+        c1: Contains all 24 single-qubit Cliffords.
```

Comment:
Isn't this a constant? Why is it being passed in?

Reply:
Are you referring to the gate sequences in c1, s1 etc? Yes, they are constants. However, they are generated outside this function so that the code doesn't have to call `_single_qubit_cliffords()` and build all the lists of gates every time it needs to generate a random two-qubit Clifford.

Reply:
You could change it to a lazily initialized module-level constant. Alternatively, pack all these arguments together into a `NamedTuple` type.